Cyclone: Cyclone Analyzer

Libraries And Opening Images

In [66]:
! pip install sympy

In [67]:
from sympy import Point, Line

In [68]:
from PIL import Image
import numpy as np
import matplotlib.pyplot as plt
import cv2

# Opening input images
# In the brackets add the path of your image
im1 = Image.open("1.jpg")
im2 = Image.open("2.jpg")
im3 = Image.open("3.jpg")

Coordinates And Calculations

In [69]:
# The coordinates have been obtained by hit and trial method which has not been shown here
# b refers to Belna and a refers to Ambali

b1 = [237,114]
a1 = [546,181]

b2 = [224,121]
a2 = [541,188]

b3 = [216,135]
a3 = [530,195]

ru = (362,499)  # ru - Reunion
m = (415,475)   # m stands for Mauritius

arnn = [200,269]    # Antsiranana Coordinates
bka = [95,617]      # Benonoka Coordinates
atvo = [141,438]    # Antananarivo Coordinates

# The actual distance between the two blue points in Img 1 is given to be 250
adb = 250

# Assume the pixel distance between Reunion and Mauritius to be PDB
pdb = ((ru[0]-m[0])**2+(ru[1]-m[1])**2)**(0.5)

# The scaling between Actual Distance and Pixel Distance (Blue - Right)
ratio1 = adb / pdb

# The actual distance between the two green points in Img 1 is given to be 1500
adg = 1500

# Assume the pixel distance between Antsiranana and Benonoka to be PDB
pdg = ((arnn[0]-bka[0])**2 + (arnn[1]-bka[1])**2)**(0.5)

# The scaling between Actual Distance and Pixel Distance (Green - Left)
ratio2 = adg / pdg

# print(ratio1)
# print(ratio2)

# Average Scaling from Ratio 1 and Ratio 2
avg_scal = (ratio1 + ratio2)/2

Apojuncture

In [70]:
# def join_line(p1, p2):
#     on = (p1[1] - p2[1])
#     zr = (p1[0] - p2[0])
#     det = (p1[0]*p2[1] - p2[0]*p1[1])
#     return on, -zr, -det


def intersect(A, B, C, D):
    # Consider Line AB as a1x + b1y = c1
    a1 = -B[1] + A[1]
    b1 = -A[0] + B[0]
    c1 = -(B[1])*(A[0]) + (B[0])*(A[1])
 
  # Consider Line CD as a2x + b2y = c2
    a2 = -D[1] + C[1]
    b2 = -C[0] + D[0]
    c2 = -(D[1])*(C[0]) + (D[0])*(C[1])
 
    det = a1*b2 - a2*b1     # Determinant
    
    if(det == 0):
        return False
    
    x = (b2*c1 - b1*c2)/det
    y = (a1*c2 - a2*c1)/det
    return x, y

In [71]:

l1 = Line(b2, b3)
l2 = Line(a2, a3)
Apojuncture = intersect(b2, b3, a2, a3)

print(Apojuncture)

(-17.306122448979593, 543.2857142857143)


In [72]:
def dist(P1, P2):
    dsq = ((P1[0]-P2[0])**2+(P1[1]-P2[1])**2)
    return np.sqrt(dsq)

Distance Between Apojuncture And Capital

In [73]:
# Pixel-wise distance between Apojuncture and Antananarivo
# Unit of Distance is Km
pd1 = dist(Apojuncture,atvo)

#Actual Distance
Dist_Apojuncture_Capital = pd1 * avg_scal

print("Distance:",Dist_Apojuncture_Capital)

Distance: 800.7461169400236


Cyclone Speeds

In [74]:
# Speed of Cyclones
# The times are shown in the image
# Unit of Speed is Km/Hr and Unit of time is Hr
t1 = 1444 - 1403
t1 = t1/60
t2 = 1539 - 1444
t2 = t2/60
bs1 = dist(b1,b2)/t1
as1 = dist(a1, a2)/t1

bs2 = dist(b2, b3)/t2
as2 = dist(a2, a3)/t2

Speed_of_Belna = (bs1+bs2)/2
Speed_of_Ambali = (as1+as2)/2

print("Speed of Belna:",Speed_of_Belna*avg_scal)
print("Speed of Ambali:",Speed_of_Ambali*avg_scal)

Speed of Belna: 66.94820587023256
Speed of Ambali: 43.85206487764847


Time Taken By The Cyclones

In [75]:
# All Time related quantities have unit as hour

time_taken_by_Belna = dist(b3,Apojuncture)/Speed_of_Belna
time_taken_by_Ambali = dist(a1,Apojuncture)/Speed_of_Ambali

print("Time taken by Belna to reach Apojuncture:",time_taken_by_Belna)
print("Time taken by Ambali to reach Apojuncture:",time_taken_by_Ambali)
print("Time interval difference:",time_taken_by_Ambali - time_taken_by_Belna)

Time taken by Belna to reach Apojuncture: 29.58347177925747
Time taken by Ambali to reach Apojuncture: 64.32618141878636
Time interval difference: 34.7427096395289


# RESULTS

Coordinates of Apojuncture = (-17.306122448979593, 543.2857142857143)
<br>Distance between Apojuncture and Capital = 800.7461169400236 Kms

Speed of Belna: 66.94820587023256 Kms/Hr
<br>Speed of Ambali: 43.85206487764847 Kms/Hr

<br>Time taken by Belna to reach Apojuncture: 29.58347177925747 Hrs
<br>Time taken by Ambali to reach Apojuncture: 64.32618141878636 Hrs
<br>Time difference: 34.7427096395289 Hrs

# CONCLUSION

As Ambali and Belna do not reach the capital at the same time, We can conclude that they will not collide